src/main/java/org/apache/sling/commons/scheduler/impl/QuartzThreadPool.java

```diff
@@ -67,8 +72,23 @@ public void setInstanceName(final String name) {
      */
     @Override
     public boolean runInThread(final Runnable job) {
-        this.executor.execute(job);
-
+        synchronized ( this.lock ) {
+            if ( this.counter == 0 ) {
+                return false;
+            }
+            this.counter--;
+        }
+        final Runnable r = () -> {
+            try {
+                job.run();
+            } finally {
+                synchronized ( this.lock ) {
+                    this.counter++;
+                    this.lock.notify();
+                }    
+            }
+        };
+        this.executor.execute(r);
```

Comment:
could go more paranoia by checking for Error or RuntimeException here (and then inc the counter as well) ...

Reply:
Maybe, although you might end up with double decrement? What if the runnable is executed, runs the finally part and then a RuntimeException is thrown?

Reply:
good point, it wouldn't be well-defined in what state the runnable is